File: frontend/types/conversation.ts
export type MessageRole = 'user' | 'assistant' | 'system' | 'tool';

export interface Message {
  id: number;
  role: MessageRole;
  content: string;
  tool_name?: string;
  tool_call_id?: string;
  created_at: string;
}

export interface Conversation {
  id: number;
  title?: string;
  created_at: string;
  updated_at: string;
  messages: Message[];
}

export interface ConversationListItem {
  id: number;
  title?: string;
  created_at: string;
  updated_at: string;
  message_count: number;
}

File: backend/app/schemas/conversation.py
from datetime import datetime

from pydantic import BaseModel

from app.models.conversation import MessageRole


class MessageBase(BaseModel):
    role: MessageRole
    content: str
    tool_name: str | None = None
    tool_call_id: str | None = None


class MessageCreate(MessageBase):
    pass


class MessageResponse(MessageBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class ConversationBase(BaseModel):
    title: str | None = None


class ConversationCreate(ConversationBase):
    pass


class ConversationResponse(ConversationBase):
    id: int
    created_at: datetime
    updated_at: datetime
    messages: list[MessageResponse] = []

    class Config:
        from_attributes = True


class ConversationListResponse(BaseModel):
    id: int
    title: str | None
    created_at: datetime
    updated_at: datetime
    message_count: int = 0

    class Config:
        from_attributes = True
